When the delegate is unavailable in `PlaceholderRenderingContextSource::setPlaceholderBuffer`, save the image buffer and propagate it later to the delegate

mbrodesser-Igalia opened the pull request:
#### 153da00a252619799ba4b32cd0ac6c5b8faf6a35
<pre>
When the delegate is unavailable in `PlaceholderRenderingContextSource::setPlaceholderBuffer`, save the image buffer and propagate it later to the delegate
<a href="https://bugs.webkit.org/show_bug.cgi?id=290042">https://bugs.webkit.org/show_bug.cgi?id=290042</a>

Reviewed by Matt Woodrow.

Ensures the image buffer is propagated to the delegate, when the former
becomes available before the latter. This happened reliably for
`canvas.2d.fillText-FontFace.html`.

Combined changes:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/manual/text/canvas.2d.fillText-FontFace-expected-mismatch.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/manual/text/canvas.2d.fillText-FontFace.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/manual/text/canvas.2d.fillText-expected-mismatch.html: Added.

* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/manual/text/canvas.2d.fillText.html:
This test passed already without the fix.
The actual issue could only reliably (dozens of times) be reproduced by
opening this test file in GTK&apos;s MiniBrowser with the mouse arrow symbol
next to the browser window. With the fix, the issue is still
reproducible.

* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/manual/text/empty-ref.html: Added.

* LayoutTests/platform/mac-wk1/TestExpectations:
Tried reproducing the issue for WK1 with GTK, but GTK seems to lack
support for WK1: <a href="https://bugs.webkit.org/show_bug.cgi?id=290445.">https://bugs.webkit.org/show_bug.cgi?id=290445.</a>

* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::commitToPlaceholderCanvas):
* Source/WebCore/html/canvas/PlaceholderRenderingContext.cpp:
(WebCore::PlaceholderRenderingContextSource::setPlaceholderBuffer):
(WebCore::PlaceholderRenderingContextSource::setContentsToLayer):
* Source/WebCore/html/canvas/PlaceholderRenderingContext.h:

Canonical link: <a href="https://commits.webkit.org/295259@main">https://commits.webkit.org/295259@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/441c35d1ae94abcac18370becc9a682b5d73d1d3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104397 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24101 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14517 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109599 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55065 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106437 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24479 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32648 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79265 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107403 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19032 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94215 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59595 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18839 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12290 "Found 2 new test failures: http/tests/security/cookie-module-propagate.html http/tests/security/cookie-module.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54427 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88523 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12342 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111983 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31554 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23241 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88303 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31918 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90459 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87969 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22441 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32868 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10634 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/26022 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31482 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36807 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31276 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34612 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32836 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebCore/html/canvas/PlaceholderRenderingContext.cpp:
```diff
@@ -111,6 +111,13 @@ void PlaceholderRenderingContext::setPlaceholderBuffer(Ref<ImageBuffer>&& buffer
+{
+    // Setting the image buffer happens in `PlaceholderRenderingContext::setPlaceholderBuffer`.
+    // That's considered a deferred operation.
+    return true;
```
It feels a bit weird pretending to have deferred operations, in order to force a `prepareCanvasesForDisplayOrFlushIfNeeded()` call (which isn't actually needed), we just also get the side effect of the rest of the rendering update happening which is needed.

Seems like we should more directly be requesting a rendering update when we set an external ImageBuffer as the content software the canvas.

Adding @kkinnunen-apple in case he has better ideas.

kkinnunen-apple replied:
When placeholder gets updated, it should cause dirty area to the layer it is contained in. That dirty area should be the signal to re-render in case the dirty is visible.

kkinnunen-apple commented on Source/WebCore/html/canvas/PlaceholderRenderingContext.cpp:
```diff
@@ -111,6 +111,13 @@ void PlaceholderRenderingContext::setPlaceholderBuffer(Ref<ImageBuffer>&& buffer
     canvasBase().setImageBufferAndMarkDirty(WTFMove(buffer));
```
IIRC the calls should end up here, and this should traverse the renderer tree of the element upwards.

mattwoodrow commented on Source/WebCore/html/canvas/PlaceholderRenderingContext.cpp:
```diff
@@ -76,6 +80,10 @@ void PlaceholderRenderingContextSource::setContentsToLayer(GraphicsLayer& layer)
     Locker locker { m_lock };
     m_delegate = layer.createAsyncContentsDisplayDelegate(m_delegate.get());
+    if (m_imageBufferForDelegate) {
+        m_delegate->tryCopyToLayer(*m_imageBufferForDelegate);
```
std::exchange?

Commit:
When the delegate is unavailable in `PlaceholderRenderingContextSource::setPlaceholderBuffer`, save the image buffer and propagate it later to the delegate
https://bugs.webkit.org/show_bug.cgi?id=290042

Reviewed by Matt Woodrow.

Ensures the image buffer is propagated to the delegate, when the former
becomes available before the latter. This happened reliably for
`canvas.2d.fillText-FontFace.html`.

Combined changes:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/manual/text/canvas.2d.fillText-FontFace-expected-mismatch.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/manual/text/canvas.2d.fillText-FontFace.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/manual/text/canvas.2d.fillText-expected-mismatch.html: Added.

* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/manual/text/canvas.2d.fillText.html:
This test passed already without the fix.
The actual issue could only reliably (dozens of times) be reproduced by
opening this test file in GTK's MiniBrowser with the mouse arrow symbol
next to the browser window. With the fix, the issue is still
reproducible.

* LayoutTests/imported/w3c/web-platform-tests/html/canvas/offscreen/manual/text/empty-ref.html: Added.

* LayoutTests/platform/mac-wk1/TestExpectations:
Tried reproducing the issue for WK1 with GTK, but GTK seems to lack
support for WK1: https://bugs.webkit.org/show_bug.cgi?id=290445.

* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::commitToPlaceholderCanvas):
* Source/WebCore/html/canvas/PlaceholderRenderingContext.cpp:
(WebCore::PlaceholderRenderingContextSource::setPlaceholderBuffer):
(WebCore::PlaceholderRenderingContextSource::setContentsToLayer):
* Source/WebCore/html/canvas/PlaceholderRenderingContext.h:

Canonical link: https://commits.webkit.org/295259@main
